sdks/java/core/src/test/java/org/apache/beam/sdk/io/range/ByteKeyRangeTrackerTest.java

```diff
@@ -28,66 +28,96 @@
+  private static final double INITIAL_RANGE_SIZE = 0x34 - 0x12;
+  private static final ByteKeyRange INITIAL_RANGE = ByteKeyRange.of(INITIAL_START_KEY, END_KEY);
+  private static final double NEW_RANGE_SIZE = 0x34 - 0x14;
+  private static final ByteKeyRange NEW_RANGE = ByteKeyRange.of(NEW_START_KEY, END_KEY);
 
   /** Tests for {@link ByteKeyRangeTracker#toString}. */
   @Test
   public void testToString() {
-    ByteKeyRangeTracker tracker = ByteKeyRangeTracker.of(RANGE);
-    String expected = String.format("ByteKeyRangeTracker{range=%s, position=null}", RANGE);
+    ByteKeyRangeTracker tracker = ByteKeyRangeTracker.of(INITIAL_RANGE);
+    String expected = String.format("ByteKeyRangeTracker{range=%s, position=null}", INITIAL_RANGE);
     assertEquals(expected, tracker.toString());
 
-    tracker.tryReturnRecordAt(true, MIDDLE_KEY);
-    expected = String.format("ByteKeyRangeTracker{range=%s, position=%s}", RANGE, MIDDLE_KEY);
+    tracker.tryReturnRecordAt(true, INITIAL_START_KEY);
+    tracker.tryReturnRecordAt(true, INITIAL_MIDDLE_KEY);
+    expected =
+        String.format("ByteKeyRangeTracker{range=%s, position=%s}", INITIAL_RANGE,
+            INITIAL_MIDDLE_KEY);
+    assertEquals(expected, tracker.toString());
+  }
+
+  /** Tests for updating the start key to the first record returned. */
+  @Test
+  public void testUpdateStartKey() {
+    ByteKeyRangeTracker tracker = ByteKeyRangeTracker.of(INITIAL_RANGE);
+
+    tracker.tryReturnRecordAt(true, NEW_START_KEY);
+    String expected =
+        String.format("ByteKeyRangeTracker{range=%s, position=%s}", NEW_RANGE, NEW_START_KEY);
     assertEquals(expected, tracker.toString());
   }
 
   /** Tests for {@link ByteKeyRangeTracker#of}. */
   @Test
   public void testBuilding() {
-    ByteKeyRangeTracker tracker = ByteKeyRangeTracker.of(RANGE);
+    ByteKeyRangeTracker tracker = ByteKeyRangeTracker.of(INITIAL_RANGE);
 
-    assertEquals(START_KEY, tracker.getStartPosition());
+    assertEquals(INITIAL_START_KEY, tracker.getStartPosition());
     assertEquals(END_KEY, tracker.getStopPosition());
   }
 
   /** Tests for {@link ByteKeyRangeTracker#getFractionConsumed()}. */
   @Test
   public void testGetFractionConsumed() {
-    ByteKeyRangeTracker tracker = ByteKeyRangeTracker.of(RANGE);
+    ByteKeyRangeTracker tracker = ByteKeyRangeTracker.of(INITIAL_RANGE);
     double delta = 0.00001;
 
     assertEquals(0.0, tracker.getFractionConsumed(), delta);
 
-    tracker.tryReturnRecordAt(true, START_KEY);
+    tracker.tryReturnRecordAt(true, INITIAL_START_KEY);
+    assertEquals(0.0, tracker.getFractionConsumed(), delta);
+
+    tracker.tryReturnRecordAt(true, INITIAL_MIDDLE_KEY);
+    assertEquals(0.5, tracker.getFractionConsumed(), delta);
+
+    tracker.tryReturnRecordAt(true, BEFORE_END_KEY);
+    assertEquals(1 - 1 / INITIAL_RANGE_SIZE, tracker.getFractionConsumed(), delta);
+
+    // Should be accurate for updated start key
+    tracker = ByteKeyRangeTracker.of(INITIAL_RANGE);
```

Comment:
please make it a separate test.

`testGetFractionConsumed` as it was, and then `testGetFractionConsumedLateStartKey`, for example